polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h

```diff
@@ -13,15 +13,103 @@
 #ifndef MLIR_DIALECT_POLYGEIST_ANALYSIS_REACHINGDEFINITIONANALYSIS_H
 #define MLIR_DIALECT_POLYGEIST_ANALYSIS_REACHINGDEFINITIONANALYSIS_H
 
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
 #include "mlir/Analysis/DataFlow/DenseAnalysis.h"
 #include "mlir/Analysis/DataFlow/SparseAnalysis.h"
+#include "mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h"
 #include "mlir/Dialect/Polygeist/Utils/AliasUtils.h"
 #include <set>
 #include <variant>
 
 namespace mlir {
 namespace polygeist {
 
+/// A class representing the value a formal argument of a function has the
+/// caller of the function.
+class UnderlyingValue {
+  friend raw_ostream &operator<<(raw_ostream &, const UnderlyingValue &);
+
+public:
+  /// Create an underlying value state with a known underlying value.
+  explicit UnderlyingValue(std::optional<Value> val = std::nullopt)
+      : val(val) {}
+
+  /// Whether the state is uninitialized.
+  bool isUninitialized() const { return !val.has_value(); }
+
+  /// Returns the underlying value.
+  Value get() const {
+    assert(!isUninitialized() && "Expecting the state to be initialized");
+    return *val;
+  }
+
+  /// Join two underlying values. If there are conflicting underlying values,
+  /// go to the pessimistic value.
+  static UnderlyingValue join(const UnderlyingValue &lhs,
+                              const UnderlyingValue &rhs) {
+    if (lhs.isUninitialized())
+      return rhs;
+    if (rhs.isUninitialized())
+      return lhs;
+    return lhs.val == rhs.val ? lhs : UnderlyingValue(Value{});
+  }
+
+  /// Compare underlying values.
+  bool operator==(const UnderlyingValue &rhs) const { return val == rhs.val; }
+
+  void print(raw_ostream &os) const { os << val; }
+
+private:
+  std::optional<Value> val;
+};
+
+struct UnderlyingValueLattice : public dataflow::Lattice<UnderlyingValue> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(UnderlyingValueLattice)
+  using Lattice::Lattice;
+};
+
+/// An analysis that uses forwarding of values along control-flow and callgraph
+/// edges to determine single underlying values for block arguments.
+class UnderlyingValueAnalysis
+    : public dataflow::SparseDataFlowAnalysis<UnderlyingValueLattice>,
+      public RequiredDataFlowAnalyses<UnderlyingValueAnalysis> {
+  friend class RequiredDataFlowAnalyses;
+
+public:
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(UnderlyingValueAnalysis)
```

Comment:
Why is this added? Given that `UnderlyingValueAnalysis` is not in an anonymous namespaces, should we use `MLIR_DECLARE_EXPLICIT_TYPE_ID` and `MLIR_DEFINE_EXPLICIT_TYPE_ID`?
(same for `UnderlyingValueLattice`)

Reply:
Not sure, I guess I should use `MLIR_DECLARE_EXPLICIT_TYPE_ID` ?